src/Raven.Client/Documents/DocumentStore.cs

```diff
@@ -158,28 +163,79 @@ public override RequestExecutor GetRequestExecutor(string database = null)
             if (database == null)
                 database = Database;
 
-            if (_requestExecutors.TryGetValue(database, out Lazy<RequestExecutor> lazy))
-                return lazy.Value;
+            Lazy<RequestExecutor> lazy;
+            lock (_requestExecutors)
+            {
+
+                if (_requestExecutors.TryGetValue(database, out lazy))
+                    return lazy.Value;
 
-            lazy = Conventions.DisableTopologyUpdates == false
-                ? new Lazy<RequestExecutor>(() =>
+                RequestExecutor CreateRequestExecutor()
                 {
-                    var requestExecutor = RequestExecutor.Create(Urls, database, Certificate, Conventions);
+                    var requestExecutor = RequestExecutor.Create(Urls, database, Certificate, Conventions, _useCompression);
                     RequestExecutorCreated?.Invoke(this, requestExecutor);
                     return requestExecutor;
-                })
-                : new Lazy<RequestExecutor>(() =>
+                }
+                
+                RequestExecutor CreateRequestExecutorForSingleNode()
                 {
-                    var forSingleNode = RequestExecutor.CreateForSingleNodeWithConfigurationUpdates(Urls[0], database, Certificate, Conventions);
+                    var forSingleNode = RequestExecutor.CreateForSingleNodeWithConfigurationUpdates(Urls[0], database, Certificate, Conventions, _useCompression);
                     RequestExecutorCreated?.Invoke(this, forSingleNode);
                     return forSingleNode;
-                });
+                }
 
-            lazy = _requestExecutors.GetOrAdd(database, lazy);
+                lazy = Conventions.DisableTopologyUpdates == false
+                    ? new Lazy<RequestExecutor>(CreateRequestExecutor)
+                    : new Lazy<RequestExecutor>(CreateRequestExecutorForSingleNode);
+
+                lazy = _requestExecutors.GetOrAdd(database, lazy);
+            }
 
             return lazy.Value;
         }
 
+        private bool _useCompression;
+
+        /// <inheritdoc />
+        public override bool UseCompression
```

Comment:
Move to Conventions? Those are frozen after DocumentStore.Initialize and cannot be changed. No need to lock anything then.

Reply:
After that most of the code will be simplified.

Reply:
Moving it to conventions will simplify code indeed. Will do so.